jplag.frontend.R/src/main/java/de/jplag/R/RTokenConstants.java

```diff
@@ -0,0 +1,41 @@
+package de.jplag.R;
+
+import de.jplag.TokenConstants;
+
+/**
+ * Tokens in R that are deemed important when comparing submissions for plagiarisms. Based on an R frontend for JPlag
+ * v2.15 by Olmo Kramer, see their <a href="https://github.com/CodeGra-de/jplag/tree/master/jplag.frontend.R">JPlag
+ * fork</a>.
+ * @author Robin Maisch
+ */
+public interface RTokenConstants extends TokenConstants {
+    int FILE_END = 0;
+    int SEPARATOR_TOKEN = 1;
+
+    int BEGIN_FUNCTION = 2;
+    int END_FUNCTION = 3;
+    int FUNCTION_CALL = 4;
+    int NUMBER = 5;
+    int STRING = 6;
+    int BOOL = 7;
+    int ASSIGN = 8;
+    int ASSIGN_FUNC = 9;
+    int ASSIGN_LIST = 10;
+    int HELP = 11;
+    int INDEX = 12;
+    int PACKAGE = 13;
+    int IF_BEGIN = 14;
+    int IF_END = 15;
+    int FOR_BEGIN = 16;
+    int FOR_END = 17;
+    int WHILE_BEGIN = 18;
+    int WHILE_END = 19;
+    int REPEAT_BEGIN = 20;
+    int REPEAT_END = 21;
+    int NEXT = 22;
+    int BREAK = 23;
+    int COMPOUND_BEGIN = 24;
+    int COMPOUND_END = 25;
+
+    int NUM_DIFF_TOKENS = 26;
```

Comment:
Currently, this frontend does not have many tokens. From your own testing, do you think R is just less expressive than e.g. java or do you think this token set should be expanded in the future?

Reply:
AFAICT, R code is mostly just a sequence of function definition containing function calls and arithmetic operations. Syntactically, there really does not seem to be very much to do, and I would say that the current token set covers the elements given by the grammar quite well.

It turns out, though, that _some_ new syntactic elements have been introduced in R v4.0.0 and up, which are not represented in the [R grammar](https://github.com/antlr/grammars-v4/blob/master/r/R.g4) in the antlr4 repository yet, e.g. lambda functions `\(x) x*x` or the pipe operator `|>`. So, it is actually the grammar that should be expanded first if we want to continue to rely on that.